### ISSUE TITLE
zipking sampling percentage configuration

### ISSUE THREAD

**Issue by golonzovsky** [opentracing-contrib/nginx-opentracing#30]
Currently there is no way (please correct me if I wrong) to specify sampling percentage for `Zipkin` tracer. This renders zipkin support not so usable on high load prod environments, as amount of traces generated is more than actual traffic. 

Would it be possible to provide sampling param same as [for eg. jaeger](https://github.com/opentracing-contrib/nginx-opentracing/blob/master/jaeger/doc/Directives.md#jaeger_sampler_param)?

**Comment by rnburn** [comment]
That's right. The zipkin [library](https://github.com/rnburn/zipkin-cpp-opentracing) doesn't provide sampling in the way jaeger does.

But I'd be happy to take a PR for it if it's something you'd like to work on.

**Comment by pingles** [comment]
I'd really like to have support for this too- @rnburn I can _try_ and take a look, do you have any suggestions about where you'd like to see support/how it would be implemented? I.e. perhaps some kind of `opentracing_sampler_probability 0.01;` type option?

Am I right in thinking that this should also set `X-B3-Sampled: 0` in the downstream tracing headers?

**Comment by rnburn** [comment]
@pingles - yes that's correct it should set that sampled header.

Additionally, it should look at the standard opentracing `sampling.priority` tag -- described [here](https://github.com/opentracing/specification/blob/master/semantic_conventions.md).

You might check out
* Jaeger's [probabilistic sampler](
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/src/jaegertracing/samplers/ProbabilisticSampler.h)
* LightStep' support for [`sampling.priority`](
https://github.com/lightstep/lightstep-tracer-cpp/pull/94)

And I think adding a `sampling_rate` field to this [structure](https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin_opentracing/include/zipkin/opentracing.h#L6) would be a good place to control the behavior. It would also need to be added to the [JSON configuration](https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin_opentracing/src/tracer_factory.cc#L8).

**Comment by pingles** [comment]
I've started taking a look at this and trying to get my head around the interplay between the different modules.

Looking at the Jaeger module, the determination of whether to propagate the sampling decision, or determine whether to sample is done inside `::StartSpanWithOptions`. Would I be correct in thinking that the right place to inject the same logic for the Zipkin lib would be it's equivalent: https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin_opentracing/src/opentracing.cc#L277

Thanks @rnburn for your patience in helping me figure this out :)

**Comment by pingles** [comment]
Ok cool, I found https://github.com/opentracing/specification/blob/master/specification.md and wasn't sure whether sampling was something that's tracer specific (and on a `SpanContext` like in the Jaeger lib, or associated to a `Span`). I've added the code to the place above, now trying to figure out why that's not propagating to the forwarded request's HTTP headers :)

**Comment by pingles** [comment]
I've started working on this and I've managed to (brutally) get it to do the right thing (at least from observing the downstream request headers):

https://github.com/pingles/zipkin-cpp-opentracing/commit/67d68c671fd0a2cad29e8444f9872dedf4961a24

I'm going to try and carry on tomorrow- I need to fix everything that's referenced in the commit message, and additionally make sure that spans are only reported if they've been sampled (missed that out of my message).

I'm certain that I've misunderstood the relationship between the classes so if anyone more familiar could have a scan and point out where I've made a mistake/should change something I'd really appreciate it. I'm by no means an expert C++ engineer either so please correct me on that also.

Thanks again for supporting me to add this.

**Comment by rnburn** [comment]
Thanks @pingles - I went through and made a few suggestions.

**Comment by pingles** [comment]
Awesome, thanks @rnburn- I'll make those changes tomorrow, appreciate the feedback.

**Comment by pingles** [comment]
I've made a few more improvements- is there a recommended way to determine whether a span is the root or not? Looking at https://github.com/opentracing/opentracing-cpp/blob/master/include/opentracing/tracer.h#L39 suggests that if `references` was empty then it could be considered the root span and we could then consult a sampler to determine whether to sample?

**Comment by pingles** [comment]
@rnburn I've copied the Jaeger implementation for determining whether a span is the root (checking whether references contains a `ChildOf` relationship. https://github.com/pingles/zipkin-cpp-opentracing/commit/d7628b250fe1991b6ce1308dbcb20040ab94ab0b

Unfortunately, it doesn't seem to work when I test it inside the nginx module- passing sample headers with curl:

```
curl -H "X-B3-Sampled: 1" -H "X-B3-SpanId: c4adf4696a1334" -H "X-B3-TraceId: c4adf4696a1334" http://localhost
```

I have a dummy app running that's fronted by nginx and prints the incoming request headers and it shows:

```TraceID: 00c4adf4696a1334
Sampled: 0
```

So... close, but not quite enough :)

I'll keep debugging to try and figure out where the missing chain is but if anyone else spots it before me please let me know! All work is in https://github.com/pingles/zipkin-cpp-opentracing/tree/sampling currently.

**Comment by pingles** [comment]
Related to the above- the other B3 headers are propagated correctly so it's either an issue with the sampled header specifically, or, something to do with my implementation of:

```c++
bool hasParent(const ot::StartSpanOptions &options) const {
    for (auto ref : options.references) {
      if (ref.first == ot::SpanReferenceType::ChildOfRef) {
        return true;
      }
    }
    return false;
  }
```

**Comment by rnburn** [comment]
@pingles - I think the parseBool function for reading the sampling header that I wrote might be the problem.
https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin_opentracing/src/propagation.cc#L34-L46
it always returns true even if sampling is suppose to be off.

(Sorry)

**Comment by rnburn** [comment]
Also, does
```
curl -H "X-B3-Sampled: 1" -H "X-B3-SpanId: c4adf4696a1334" -H "X-B3-TraceId: c4adf4696a1334" http://localhost
```
have a space in front of the header values? that might not work for propagation.

**Comment by pingles** [comment]
@rnburn Ah... so `parseBool` returns whether the value could be parsed?

```
bool sampled;
if (!parseBool(value, sampled)) {
  return ot::make_unexpected(ot::span_context_corrupted_error);
}
flags |= sampled_flag;
```

So that should change to:

```
if sampled {
  flags |= sampled_flag;
}
```

**Comment by rnburn** [comment]
Yes, that looks right. I think it's also hard-coded for inject: https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin_opentracing/src/propagation.cc#L67

**Comment by pingles** [comment]
Awesome, I’ll try both of those a little bit later and let you know how I
get on.

**Comment by pingles** [comment]
I've also just seen that `OtSpan` has nothing relating to sampling (but does copy other items like trace IDs etc.). It doesn't look like that has anything related to passing context along but does seem to be involved in sending the trace data?

**Comment by pingles** [comment]
I'm an idiot- just seen there's some nice tests to build on, will start there to figure this out.

**Comment by rnburn** [comment]
For the souldSample function, you don't want to be doing initializing every time like [this](https://github.com/pingles/zipkin-cpp-opentracing/blob/438942bf566813131812a374639b39413784feb5/zipkin_opentracing/src/sampling.cc#L7-L9) -- that's going to be slow (see the notes in the [example code](http://en.cppreference.com/w/cpp/numeric/random/random_device) here).

Instead could you add a function to utility.cc and reuse this [rand_source](http://en.cppreference.com/w/cpp/numeric/random/random_device)? (you can move the variable to be static to the translation unit instead of the function).

**Comment by pingles** [comment]
@rnburn thanks. Is it fair to store them as members instead- and then it's only initialised when the sampler is contructed (when the tracer is)?

```
public:
ProbabilisticSampler(double sample_rate) : dist_(sample_rate) {
  std::random_device device;
  rng_.seed(device());
};
private:
  std::mt19937 rng_;
  std::bernoulli_distribution dist_;
```

Or is it substantially better to use static? Also- is there any concern around thread-safe concurrent requests for numbers from the distribution, does it need to be guarded with some kind of mutex?

**Comment by rnburn** [comment]
Yeah, that's what's done here: https://github.com/rnburn/zipkin-cpp-opentracing/blob/master/zipkin/src/utility.cc.

I was thinking add something like randomBool(double p) so that you can reuse that same rand_source variable that's already initialized.

But I think what you have is ok too.

**Comment by pingles** [comment]
Ok cool, I'll go with that for now :-) I think there's two things left to do: 1) make sure that spans aren't reported unless they're sampled, 2) figure out why the randomisation doesn't seem to work from the nginx module (https://github.com/pingles/nginx-opentracing/commit/e7b586c02c7945218005a759fefc85da19158093).

Current behaviour is that whatever sampling decision is forwarded from the client is propagated correctly. However, without any existing context headers the tracer always sets spans to be unsampled. 

So very nearly there :-)

**Comment by pingles** [comment]
Oh, and check the `sampling.priority` tag that you referenced in a comment above, so 3 things.

**Comment by pingles** [comment]
I've updated the lib to only report sampled spans and am now debugging the issue with all spans being unsampled when used by the nginx module.

I've turned on the debug log and can see the following (with `opentracing_trace_locations off`):

```
2018/05/21 12:28:54 [debug] 23125#0: *1 http proxy header: "x-b3-traceid: 0948372d22ebf708"
2018/05/21 12:28:54 [debug] 23125#0: *1 http proxy header: "x-b3-spanid: 3bcb59253435d447"
2018/05/21 12:28:54 [debug] 23125#0: *1 http proxy header: "x-b3-sampled: 0"
2018/05/21 12:28:54 [debug] 23125#0: *1 http proxy header: "x-b3-parentspanid: 0000000000000000"
2018/05/21 12:28:54 [debug] 23125#0: *1 http proxy header: "x-b3-flags: 0"
```

It feels like there's something I've misunderstood in the way that spans are constructed around the request- that either the sampler is never called, or that it's not randomly drawing from the distribution. I don't suppose you have any good suggestions on where to look @rnburn please before I start debug logging everything

**Comment by pingles** [comment]
Note to myself: it feels like it might be a case of checking whether a parent span is valid, not just that it exists: https://github.com/jaegertracing/jaeger-client-cpp/blob/master/src/jaegertracing/SpanContext.h#L150

**Comment by pingles** [comment]
I _think_ everything in the underlying zipkin opentracing lib is now updated to support probabilistic sampling from within the nginx module (aside from propagating information through the tags also).

However, I'm still having issues using it within the nginx module: currently it looks like the sampling rate is incorrectly parsed- so ~1/10 requests are sampled when the sampling probability ought to be 1/1. 

I think I'll try and get the tag stuff completed on the zipkin lib and then open a PR as I'm pretty sure that is good to go- it'd be good to get some more feedback on the code to make sure I've not missed/misunderstood anything. After that I'll get back on trying to figure out what's happening within the nginx module.

**Comment by rnburn** [comment]
thanks @pavolloffay - let me know when you're ready, and I can take another look at it.

you might open up the PR, even if you're not finished - it will make it easier to review.

**Comment by pingles** [comment]
Perfect- I'll just do that then

**Comment by pingles** [comment]
I've just created the PR: https://github.com/rnburn/zipkin-cpp-opentracing/pull/20

**Comment by rnburn** [comment]
Fixed with #38 (thank you @pingles )
